pennylane/gradients/hadamard_gradient.py

```diff
@@ -309,218 +290,122 @@ def hadamard_grad(
     # Validate or get default for aux_wire
     aux_wire = _get_aux_wire(aux_wire, tape, device_wires)
 
-    g_tapes, processing_fn = _expval_hadamard_grad(tape, argnum, aux_wire)
-
-    return g_tapes, processing_fn
+    return _expval_hadamard_grad(tape, argnum, aux_wire)
 
 
 def _expval_hadamard_grad(tape, argnum, aux_wire):
     r"""Compute the Hadamard test gradient of a tape that returns an expectation value (probabilities are expectations
     values) with respect to a given set of all trainable gate parameters.
     The auxiliary wire is the wire which is used to apply the Hadamard gates and controlled gates.
     """
-    # pylint: disable=too-many-statements
     argnums = argnum or tape.trainable_params
     g_tapes = []
     coeffs = []
+    generators_per_parameter = []
 
-    gradient_data = []
     for trainable_param_idx, _ in enumerate(tape.trainable_params):
         if trainable_param_idx not in argnums:
             # parameter has zero gradient
-            gradient_data.append(0)
+            generators_per_parameter.append(0)
             continue
 
-        trainable_op, idx, p_idx = tape.get_operation(trainable_param_idx)
+        trainable_op, idx, _ = tape.get_operation(trainable_param_idx)
 
         ops_to_trainable_op = tape.operations[: idx + 1]
         ops_after_trainable_op = tape.operations[idx + 1 :]
 
         # Get a generator and coefficients
-        sub_coeffs, generators = _get_generators(trainable_op)
+        sub_coeffs, generators = _get_pauli_generators(trainable_op)
         coeffs.extend(sub_coeffs)
-
-        num_tape = 0
+        generators_per_parameter.append(len(generators))
 
         for gen in generators:
-            if isinstance(trainable_op, qml.Rot):
```

Comment:
I saw no reason to continue "natively" differenitating `Rot`.  Decomposing it first is much cleaner in terms of code and abstraction, and consumes basically the same resources. It's actually much easier to inspect the gradient tapes too.